src/particle_filter/src/particleFilter6DOF.cpp

```diff
@@ -0,0 +1,359 @@
+      }
+      b_Xpre[0][k] /= N;
+      for (int j = 0; j < N; j++) {
+	b_Xpre[1][k] += SQ(X0[j][k] - b_Xpre[0][k]);
+      }
+      b_Xpre[1][k] = sqrt(b_Xpre[1][k] / N);
+    }
+  }
+  bool iffar = updateParticles(X_1, X0, X, obs, cube, idx_obs, dist_transform, N, R, Xstd_ob, Xstd_tran);
+  if (firstObs)
+    {
+      firstObs = false;
+      memcpy(X0, X, N*sizeof(cspace));
+      memcpy(X_1, X, N*sizeof(cspace));
+    }
+  else if (iffar == true)
+    {
+      memcpy(X0, X_1, N*sizeof(cspace));
+    }
+  //calcWeight(W, N, Xstd_tran, X0, X);
+  memcpy(X_1, X0, N*sizeof(cspace));
+  //resampleParticles(X0, X, W, N);
+  memcpy(X0, X, N*sizeof(cspace));
+
+  for (int k = 0; k < cdim; k++) {
+    X_est[k] = 0;
+    for (int j = 0; j < N; j++) {
+      X_est[k] += X0[j][k];
+    }
+    X_est[k] /= N;
+  }
+  X_est_stat[1] = 0;
+  for (int j = 0; j < N; j++)
+    {
+      for (int k = 0; k < cdim; k++) {
+	X_est_stat[1] += SQ(X0[j][k] - X_est[k]);
+      }
+    }
+  X_est_stat[1] = sqrt(X_est_stat[1] / N);
+
+  if (X_est_stat[1] < 0.005 && (abs(X_est[0] - b_Xpre[0][0])>0.001 ||
+				abs(X_est[1] - b_Xpre[0][1])>0.001 ||
+				abs(X_est[2] - b_Xpre[0][2])>0.001 ||
+				abs(X_est[3] - b_Xpre[0][3])>0.001 ||
+				abs(X_est[4] - b_Xpre[0][4]) > 0.001 ||
+				abs(X_est[5] - b_Xpre[0][5]) > 0.001))
+    Xstd_scatter = 0.01;
+  else
+    Xstd_scatter = 0.0001;
+}
+
+void particleFilter::createParticles(cspace *X, cspace b_Xprior[2],
+				     int n_particles)
+{
+  random_device rd;
+  mt19937 e2(rd());
+  normal_distribution<double> dist(0, 1);
+  int cdim = sizeof(cspace) / sizeof(double);
+  for (int i = 0; i < n_particles; i++){
+    for (int j = 0; j < cdim; j++){
+      X[i][j] = b_Xprior[0][j] + b_Xprior[1][j] * (dist(e2));
+    }
+  }
+}
+
+bool particleFilter::updateParticles(cspace *X_1, cspace *X0, cspace *X, double cur_M[3], 
+				     double cube[3], int idx_Measure, distanceTransform *dist_transform, int n_particles,
+				     double R, double Xstd_ob, double Xstd_tran)
+{
+  random_device rd;
+  mt19937 e2(rd());
+  normal_distribution<double> dist(0, 1);
+  uniform_real_distribution<double> distribution(0, n_particles);
+  int cdim = sizeof(cspace) / sizeof(double);
+  int i = 0;
+  int count = 0;
+  bool iffar = false;
+  cspace *b_X = X0;
+  int dir = idx_Measure % 3;
+  int idx = 0;
+  double tempState[6];
+  double D;
+  double rotation[3][3];
+  double tempRot[3][3];
+  double invRot[3][3];
+  double cur_inv_M[3];
+  double tempM[3];
+  int num_Mean = 1000;
+  //cspace *sampleConfig = new cspace[num_Mean];
+  cspace meanConfig = { 0, 0, 0, 0, 0, 0 };
+  for (int t = 0; t < num_Mean; t++) {
+    int index = int(floor(distribution(e2)));
+    //memcpy(sampleConfig[t], b_X[index], sizeof(cspace));
+    for (int m = 0; m < cdim; m++) {
+      meanConfig[m] += b_X[index][m] / num_Mean;
+    }
+  }
+  // inverse-transform using sampled configuration
+  double rotationC[3][3] = { { cos(meanConfig[cdim - 1]), -sin(meanConfig[cdim - 1]), 0 },
```

Comment:
ros tf library or eigen will take care of all this matrix stuff.
In this case it is easier, more clear, and more reliable to use the existing library